### PR TITLE
[APB-934] [JR,RV] Backend service for OPRA reference generation

### PR DESCRIPTION
**Pull request by johnrevill**

### CODE REVIEW

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      case (Valid(_), Valid(_)) => Valid(())
+      case (Invalid(f1),Invalid(f2)) => Invalid(Failure(f1.errors ++ f2.errors))
+      case (Valid(_), f@Invalid(_)) => f
+      case (f@Invalid(_), Valid(_)) => f
+    }
+  }
+}
+
+@Singleton
+class AgentEpayeRegistrationValidator @Inject()() {
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Is empty @Inject()() needed here?

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  "uk.gov.hmrc" %% "play-auditing" % "2.10.0",
+  "uk.gov.hmrc" %% "play-auth" % "1.1.0",
+  "uk.gov.hmrc" %% "play-config" % "4.3.0",
+  "uk.gov.hmrc" %% "play-graphite" % "3.2.0",
+  "uk.gov.hmrc" %% "play-health" % "2.1.0",
+  "uk.gov.hmrc" %% "logback-json-logger" % "3.1.0",
+  "de.threedimensions" %% "metrics-play" % "2.5.13",
+  "uk.gov.hmrc" %% "play-reactivemongo" % "5.2.0",
+  "uk.gov.hmrc" %% "microservice-bootstrap" % "5.16.0",
+  "org.typelevel" %% "cats" % "0.9.0"
```
Comment (arturopala) [inline build.sbt]: As first major cats version is in the pipeline maybe we should try to catch api changes early? [cats 1.0-MF](https://github.com/typelevel/cats/releases/tag/v1.0.0-MF)

File: conf/app.routes [hunk 3]
```diff
@@ -1,3 +1,3 @@
 # microservice specific routes
 
-GET        /hello-world             uk.gov.hmrc.agentepayeregistration.controllers.MicroserviceHelloWorld.hello
+POST        /register             @uk.gov.hmrc.agentepayeregistration.controllers.AgentEpayeRegistrationController.register()
```
Comment (arturopala) [inline conf/app.routes]: REST resource names should be plural nouns, not verbs: [RestFul API Conventions](https://confluence.tools.tax.service.gov.uk/display/AR/RestFul+API+Conventions)

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  val initialAgentReference = "HX2000"
+
+  def create(request: RegistrationRequest)(implicit ec: ExecutionContext): Future[AgentReference] = {
+    val mongoCodeDuplicateKey = 11000
+
+    for {
+      maybeRegDetails <- collection.find(obj()).sort(obj("agentReference" -> -1)).one[RegistrationDetails]
+      nextAgentRef = maybeRegDetails match {
+        case Some(regDetails) => regDetails.agentReference.newReference
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: This algorithm is highly insecure considering race-condition, maybe we should follow advice here: https://docs.mongodb.com/v3.0/tutorial/create-an-auto-incrementing-field/

File: it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import play.api.test.{FakeHeaders, FakeRequest}
+import uk.gov.hmrc.agentepayeregistration.repository.AgentEpayeRegistrationRepository
+
+class AgentEpayeRegistrationControllerISpec extends BaseControllerISpec {
+  private lazy val controller: AgentEpayeRegistrationController = app.injector.instanceOf[AgentEpayeRegistrationController]
+  private lazy val repo = app.injector.instanceOf[AgentEpayeRegistrationRepository]
+
+  "submitting valid details to /register" should {
+    "respond with HTTP 200 with a the new unique PAYE code in the response body" in {
+      val postData = Json.parse(
```
Comment (ZDevelop94) [inline it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala]: postData can be refactored to only be declared once, with methods that remove or add fields. Can look up the methods if you want.

File: build.sbt (hunk 2 above)
Comment (rvijayan) [inline build.sbt]: I have read the release notes and we don't have any breaking changes at the moment. We can easily migrate to ```1.0.0``` when it's out.

File: it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import play.api.libs.json.Json
+import play.api.test.Helpers._
+import play.api.test.{FakeHeaders, FakeRequest}
+import uk.gov.hmrc.agentepayeregistration.repository.AgentEpayeRegistrationRepository
+
+class AgentEpayeRegistrationControllerISpec extends BaseControllerISpec {
+  private lazy val controller: AgentEpayeRegistrationController = app.injector.instanceOf[AgentEpayeRegistrationController]
+  private lazy val repo = app.injector.instanceOf[AgentEpayeRegistrationRepository]
+
+  "submitting valid details to /register" should {
```
Comment (arturopala) [inline it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala]: update resource name

File: it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    "respond with HTTP 200 with a the new unique PAYE code in the response body" in {
+      val postData = Json.parse(
+        """
+          |{
+          |   "agentName" : "InvalidName",
```
Comment (arturopala) [inline it/uk/gov/hmrc/agentepayeregistration/controllers/AgentEpayeRegistrationControllerISpec.scala]: why the value is "Invalid..." here?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      Invalid(Failure("INVALID_FIELD", s"The $propertyName field is not an integer"))
+
+  private def isEmailAddress(field: String)(propertyName: String) =
+    if (field.matches("""(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"""))
+      Valid(())
+    else
+      Invalid(Failure("INVALID_FIELD", s"The $propertyName field is not a valid email"))
+
+  private def isPostcode(field: String)(propertyName: String) =
+    if (field.matches("^[A-Z]{1,2}[0-9][0-9A-Z]?\\s?[0-9][A-Z]{2}$|BFPO\\s?[0-9]{1,5}$"))
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Are we sure only UK postcodes should be supported? Story doesn't require this.

File: app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        case None => AgentReference(initialAgentReference)
+      }
+      _ <- insert(RegistrationDetails(nextAgentRef, request)) recover {
+        case error: DatabaseException if error.code == Some(mongoCodeDuplicateKey) => create(request)
```
Comment (ZDevelop94) [inline app/uk/gov/hmrc/agentepayeregistration/repository/AgentEpayeRegistrationRepository.scala]: Could replace error.code == Some(mongoCodeDuplicateKey) with error.code.contains(mongoCodeDuplicateKey)

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    else
+      Invalid(Failure("MISSING_FIELD", s"The $propertyName field is mandatory"))
+
+  private def isValidCharacters(field: String)(propertyName: String) =
+    if (field.matches("[a-zA-Z0-9,.()\\-\\!@\\s]+"))
+      Valid(())
+    else
+      Invalid(Failure("INVALID_FIELD", s"The $propertyName field contains invalid characters"))
+
+  private def maxLength(field: String, maxLength: Int)(propertyName: String) =
```
Comment (ZDevelop94) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Why have you opted out of using the easier to read (field: String, maxLength: Int , propertyName: String) as a way to pass in parameters?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      isInteger(field)(propertyName).andThen(_ => maxLength(field, limit)(propertyName))
+
+    def validCharsWithLimit(field: String, propertyName: String, limit: Int) =
+      isValidCharacters(field)(propertyName).andThen(_ => maxLength(field, limit)(propertyName))
+
+    def emailValidatorWithLimit(field: String, propertyName: String, limit: Int) =
+      isEmailAddress(field)(propertyName).andThen(_ => maxLength(field, limit)(propertyName))
+
+    val optionalFieldValidators = Seq(
+      request.telephoneNumber.map(x => numberWithLimit(x, "telephone number", 35)),
```
Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: what if someone wants to enter number with country prefix, like +44... ?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 8 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Yeah, at present we are only allowing UK postcodes.

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 11 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: The validation on telephone numbers wont allow these prefix as per the specifications. The validators only allow numbers. (Clarified with Nick)

Comment (arturopala) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: maybe we should expand `+` at the start before validation, do our other APIs allows prefixed phone numbers?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 10 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: The idea is to write functions which can be combined together by omitting the property name and we can add the property name later.
eg: ```maxLength("a", 100) _```

File: test/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidatorSpec.scala [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    Some("0123456789"),
+    Some("0123456780"),
+    Some("email@test.com"),
+    address)
+
+  private def anError(errorCode: String, errorMsg: String) = Invalid(Failure(Set(ValidationError(errorCode, errorMsg))))
+
+  private def padField(chars: Int) = Seq.fill(chars)('1').mkString
+
+  "validate should pass" when {
```
Comment (arturopala) [inline test/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidatorSpec.scala]: shall we test primitive validators separately?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 11 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Agent subscription uses different regex (```^[0-9- +()#x ]{0,24}$```) for tel. numbers. We are going to use the same in this service.

File: test/uk/gov/hmrc/agentepayeregistration/model/AgentReferenceSpec.scala [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+class AgentReferenceSpec extends UnitSpec {
+  "Validation on construction of an agent reference" should {
+    "disallow an empty string" in {
+      assertThrows[IllegalArgumentException] {
+        AgentReference("")
+      }
+    }
+
+    "disallow string of length other than 6" in {
```
Comment (ZDevelop94) [inline test/uk/gov/hmrc/agentepayeregistration/model/AgentReferenceSpec.scala]: could maybe word differently? Not 100% confident in what it means

File: app/uk/gov/hmrc/agentepayeregistration/models/RegistrationDetails.scala [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+import play.api.libs.functional.syntax._
+import play.api.libs.json.{Format, JsPath}
+
+case class RegistrationDetails(agentReference: AgentReference,
+                               registration: RegistrationRequest)
+
+object RegistrationDetails {
+
+  implicit val registrationDetailsFormat: Format[RegistrationDetails] = (
```
Comment (cavster) [inline app/uk/gov/hmrc/agentepayeregistration/models/RegistrationDetails.scala]: Is there a reason you not using the default play json thing for a case class?

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 15]
```diff
@@ -0,0 +1,98 @@
+package uk.gov.hmrc.agentepayeregistration.validators
+
+import cats.Semigroup
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import uk.gov.hmrc.agentepayeregistration.models.{Failure, RegistrationRequest}
+
+object ValidatedSemigroup {
+  implicit def validatedSemigroup[A] = new Semigroup[Validated[Failure, Unit]] {
+    def combine(x: Validated[Failure, Unit], y: Validated[Failure, Unit]): Validated[Failure, Unit] = (x, y) match {
```
Comment (cavster) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Could use the |@| thing to combine instead check out Francis pr on agent-subscription-frontend

File: app/uk/gov/hmrc/agentepayeregistration/models/RegistrationDetails.scala (hunk 14 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/models/RegistrationDetails.scala]: Yes. We wanted to insert a flat json structure to mongo db.

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala (hunk 15 above)
Comment (rvijayan) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: In this case, we are implementing a ```Semigroup``` for validation which allows us to perform various operations supporting Semigroup. I think ```|@|``` just combines all the results from ```Validated``` together only.

File: app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    if (field.trim.length <= maxLength)
+      Valid(())
+    else
+      Invalid(Failure("INVALID_FIELD", s"The $propertyName field exceeds $maxLength characters"))
+
+  private def isInteger(field: String)(propertyName: String) =
+    if (field.matches("[0-9]+"))
+      Valid(())
+    else
+      Invalid(Failure("INVALID_FIELD", s"The $propertyName field is not an integer"))
```
Comment (ZDevelop94) [inline app/uk/gov/hmrc/agentepayeregistration/validators/AgentEpayeRegistrationValidator.scala]: Shouldn't the validation messages match those in the Acceptance criteria i.e. 

D: Field only contains characters allowed by the standard HMRC regex for email addresses | "Sorry, the email address entered contains invalid characters"
-- | --

**Pull request merged**